framework/data_types/ndarray.h

```diff
@@ -48,341 +50,147 @@ class NDArray
    */
   template <typename D>
   explicit NDArray(const std::vector<D>& dims)
-    : rank_(dims.size()), dimensions_(rank_), strides_(rank_), size_(0), base_(nullptr)
   {
-    static_assert(std::is_integral<D>::value,
-                  "NDArray dims argument must have vector of integral types.");
-
-    // Populate dimensions
-    for (size_t i = 0; i < rank_; ++i)
-      dimensions_[i] = dims[i];
-
-    // Populate strides and size
-    size_ = 1;
-    for (size_t i = 0; i < rank_; ++i)
-    {
-      assert(dimensions_[i] > 0);
-      size_ *= dimensions_[i];
-      strides_[i] = 1;
-      for (size_t j = i + 1; j < rank_; ++j)
-        strides_[i] *= dimensions_[j];
-    }
-
-    base_ = new T[size_];
-    for (size_t i = 0; i < size_; ++i)
-      base_[i] = 0.0;
+    set_dimensions(dims);
+    initialize();
   }
 
   /**
    * Creates an array with the specified number of elements in each dimension,
-   * from an array.
+   * from a vector-list and initializes the array.
    *
-   * \param dims `std::array` list of the number of elements in each
+   * \param dims `std::vector` list of the number of elements in each
    *             dimension.
+   * \param value Initial element value.
    * \throw std::bad_alloc if memory allocation fails.
    *
    * This constructor creates an array with the specified size.
    */
-  template <typename D, size_t N>
-  explicit NDArray(const std::array<D, N>& dims)
-    : rank_(N), dimensions_(N), strides_(N), size_(0), base_(nullptr)
+  template <typename D>
```

Comment:
Is there a real purpose to templating essentially on different `int` types here?

Reply:
IIRC when I was playing with this class, this allows you to initialize  the `NDArray` with code like this: `NDArray<Real> array({5, 10});`. Here the dimensions are treated as `int`s. We don't always have the dimensions in the `size_t` type, so this way the compiler will do the right conversion for us.